esmvalcore/preprocessor/_regrid.py

```diff
@@ -797,22 +797,19 @@ def _vertical_interpolate(cube, src_levels, levels, interpolation,
                                               cube.coord_dims(src_levels))
 
     # force mask onto data as nan's
-    cube.data = da.ma.filled(cube.core_data(), np.nan)
+    npx = get_array_module(cube.core_data())
+    data = npx.ma.filled(cube.core_data(), np.nan)
 
     # Now perform the actual vertical interpolation.
     new_data = stratify.interpolate(levels,
                                     src_levels_broadcast,
-                                    cube.core_data(),
+                                    data,
                                     axis=z_axis,
                                     interpolation=interpolation,
                                     extrapolation=extrapolation)
 
     # Calculate the mask based on the any NaN values in the interpolated data.
-    mask = np.isnan(new_data)
-
-    if np.any(mask):
-        # Ensure that the data is masked appropriately.
-        new_data = np.ma.array(new_data, mask=mask, fill_value=_MDI)
+    new_data = npx.ma.masked_where(npx.isnan(new_data), new_data)
```

Comment:
this bit right here gave us a tonne of headaches back in the golden days - so let's make sure it doesn't break again (remember the tests were passing!) - do we not need to specify a fill value? Will that return a masked array even if the data is not masked at all?

Reply:
> do we not need to specify a fill value? 

The fill value is something that you specify when saving the NetCDF file:
https://github.com/ESMValGroup/ESMValCore/blob/64f7c0f52f6182cd95a893f6c5522d5d841d270c/esmvalcore/preprocessor/_io.py#L326-L333
there is no need to keep track of it through all computations.

> Will that return a masked array even if the data is not masked at all?

`new_data` is never masked because `python-stratify` only supports plain numpy arrays, that's why there is this strategy of converting any mask to `np.nan` before calling the function. The result of this line is a masked array, though `numpy` will automatically collapse that to a single value if no data is masked, e.g.:
```python
new_data = np.array([1, 2.])
np.ma.masked_where(np.isnan(new_data), new_data)
```
results in
```
masked_array(data=[1., 2.],
             mask=False,
       fill_value=1e+20)
```
and so would dask if you were to call `compute` on it.

Reply:
right on! OK - fill values are always massive thorns, and felt I needed to double check. Good man!